In [1]:
from openai import OpenAI
client = OpenAI(
    api_key="sk-****"
)
OPENAI_API_KEY = "sk-****"

In [3]:
# Example OpenAI Python library request
MODEL = "gpt-3.5-turbo"
response = client.chat.completions.create(
    model=MODEL,
    messages=[
        {"role": "system", "content": "You are a helpful assistant."},
        {"role": "user", "content": "Knock knock."},
        {"role": "assistant", "content": "Who's there?"},
        {"role": "user", "content": "Orange."},
    ],
    temperature=0,
)

response


ChatCompletion(id='chatcmpl-8TkwFjy2sRYpeUedbbphgIW1cdYcC', choices=[Choice(finish_reason='stop', index=0, message=ChatCompletionMessage(content='Orange who?', role='assistant', function_call=None, tool_calls=None))], created=1702102635, model='gpt-3.5-turbo-0613', object='chat.completion', system_fingerprint=None, usage=CompletionUsage(completion_tokens=3, prompt_tokens=35, total_tokens=38))

# ChatCompletion Object Description

- **id**: `chatcmpl-8TkwFjy2sRYpeUedbbphgIW1cdYcC`
  - 고유 식별자로, 특정 채팅 완료 사례를 식별하는 데 사용됩니다.
  
- **choices**: 
  - 각 `Choice`는 모델의 응답과 관련된 정보를 포함합니다.
    - **finish_reason**: 'stop'
      - 응답 생성이 종료된 이유를 나타냅니다.
    - **index**: 0
      - 해당 응답이 여러 응답 중 몇 번째인지 나타내는 인덱스입니다.
    - **message**: 'Orange who?'
      - 실제 모델이 생성한 메시지 내용입니다.
    - **role**: 'assistant'
      - 메시지를 생성한 역할을 나타냅니다.

- **created**: 1702102635
  - 채팅 완료가 생성된 시간을 유닉스 타임스탬프 형태로 나타냅니다.

- **model**: 'gpt-3.5-turbo-0613'
  - 사용된 모델의 버전을 나타냅니다.

- **object**: 'chat.completion'
  - 객체의 유형을 나타냅니다.

- **system_fingerprint**: 
  - 시스템 지문에 대한 정보를 포함할 수 있습니다.

- **usage**: 
  - 토큰 사용량에 대한 정보를 포함합니다.
    - **completion_tokens**: 3
    - **prompt_tokens**: 35
    - **total_tokens**: 38


In [5]:
response.choices[0].message.content


'Orange who?'

In [6]:
# example with a system message
response = client.chat.completions.create(
    model=MODEL,
    messages=[
        {"role": "system", "content": "You are a helpful assistant."},
        {"role": "user", "content": "Explain asynchronous programming in the style of the pirate Blackbeard."},
    ],
    temperature=0,
)

print(response.choices[0].message.content)


Arr, me matey! Let me tell ye a tale of asynchronous programming, in the style of the fearsome pirate Blackbeard!

Picture this, me hearties. In the vast ocean of programming, there be times when ye need to perform multiple tasks at once. But fear not, for asynchronous programming be here to save the day!

Ye see, in traditional programming, ye be waitin' for one task to be done before movin' on to the next. But with asynchronous programming, ye can be takin' care of multiple tasks at the same time, like a true pirate multitasker!

Instead of waitin' for a task to be completed, ye be sendin' it off on its own journey, while ye move on to the next task. It be like sendin' yer crewmates on different missions, each one doin' their own thing without waitin' for the others.

But how does it work, ye ask? Well, in the world of programming, ye be usin' something called callbacks or promises. These be like secret messages ye be sendin' to yer crewmates, tellin' them what to do when they be fin

# 시스템 메시지


시스템 메시지는 비서에게 다양한 성격이나 행동을 준비하는 데 사용될 수 있습니다.


는 일반적으로 또는 gpt-3.5-turbo-0301처럼 시스템 메시지에 많은 주의를 기울이지 않습니다 . 따라서 의 경우 대신 사용자 메시지에 중요한 지침을 배치하는 것이 좋습니다. 일부 개발자는 대화가 길어질수록 모델의 주의가 다른 곳으로 흩어지는 것을 방지하기 위해 대화가 끝날 무렵 시스템 메시지를 지속적으로 이동하는 데 성공했습니다.gpt-4-0314gpt-3.5-turbo-0613gpt-3.5-turbo-0301

In [7]:
# An example of a system message that primes the assistant to explain concepts in great depth
response = client.chat.completions.create(
    model=MODEL,
    messages=[
        {"role": "system", "content": "You are a friendly and helpful teaching assistant. You explain concepts in great depth using simple terms, and you give examples to help people learn. At the end of each explanation, you ask a question to check for understanding"},
        {"role": "user", "content": "Can you explain how fractions work?"},
    ],
    temperature=0,
)

print(response.choices[0].message.content)


Of course! Fractions are a way to represent parts of a whole. They are made up of two numbers: a numerator and a denominator. The numerator tells you how many parts you have, and the denominator tells you how many equal parts make up the whole.

Let's take an example to understand this better. Imagine you have a pizza that is divided into 8 equal slices. If you eat 3 slices, you can represent that as the fraction 3/8. Here, the numerator is 3 because you ate 3 slices, and the denominator is 8 because the whole pizza is divided into 8 slices.

Fractions can also be used to represent numbers less than 1. For example, if you eat half of a pizza, you can write it as 1/2. Here, the numerator is 1 because you ate one slice, and the denominator is 2 because the whole pizza is divided into 2 equal parts.

Now, let's talk about equivalent fractions. Equivalent fractions are different fractions that represent the same amount. For example, 1/2 and 2/4 are equivalent fractions because they both re

In [8]:
# An example of a system message that primes the assistant to give brief, to-the-point answers
response = client.chat.completions.create(
    model=MODEL,
    messages=[
        # {"role": "system", "content": "You are a laconic assistant. You reply with brief, to-the-point answers with no elaboration."},
        {"role": "system", "content": "당신은 간결한 조수입니다. 짧고 핵심적인 대답으로만 답변하며, 추가 설명은 하지 않습니다."},
        {"role": "user", "content": "Can you explain how fractions work?"},
    ],
    temperature=0,
)

print(response.choices[0].message.content)


Fractions represent a part of a whole or a ratio between two numbers. They consist of a numerator (top number) and a denominator (bottom number), separated by a slash (/). The numerator represents the number of parts we have, while the denominator represents the total number of equal parts in the whole.


# 퓨샷 프롬프트

어떤 경우에는 모델에게 원하는 것을 말하는 것보다 원하는 것을 모델에 보여주는 것이 더 쉽습니다.


원하는 것을 모델에 표시하는 한 가지 방법은 가짜 예제 메시지를 사용하는 것입니다.


예를 들어:

In [9]:
# The business jargon translation example, but with example names for the example messages
response = client.chat.completions.create(
    model=MODEL,
    messages=[
        {"role": "system", "content": "You are a helpful, pattern-following assistant that translates corporate jargon into plain English."},
        {"role": "system", "name":"example_user", "content": "New synergies will help drive top-line growth."},
        {"role": "system", "name": "example_assistant", "content": "Things working well together will increase revenue."},
        {"role": "system", "name":"example_user", "content": "Let's circle back when we have more bandwidth to touch base on opportunities for increased leverage."},
        {"role": "system", "name": "example_assistant", "content": "Let's talk later when we're less busy about how to do better."},
        {"role": "user", "content": "This late pivot means we don't have time to boil the ocean for the client deliverable."},
    ],
    temperature=0,
)

print(response.choices[0].message.content)


This sudden change in direction means we don't have enough time to complete the entire project for the client.


# 4. 토큰 계산

요청을 제출하면 API는 메시지를 일련의 토큰으로 변환합니다.


사용되는 토큰 수는 다음에 영향을 미칩니다.


요청 비용
응답을 생성하는 데 걸리는 시간
gpt-3.5-turbo최대 토큰 제한(4,096개 또는 8,192개 gpt-4) 에 도달하여 응답이 잘리는 경우
다음 함수를 사용하여 메시지 목록에서 사용할 토큰 수를 계산할 수 있습니다.


메시지에서 토큰을 계산하는 정확한 방법은 모델마다 다를 수 있습니다. 시간을 초월한 보장이 아닌 추정치 아래 함수의 개수를 고려하세요.


특히 선택적 함수 입력을 사용하는 요청은 아래 계산된 추정치 외에 추가 토큰을 소비합니다.

In [11]:
import tiktoken


def num_tokens_from_messages(messages, model="gpt-3.5-turbo-0613"):
    """Return the number of tokens used by a list of messages."""
    try:
        encoding = tiktoken.encoding_for_model(model)
    except KeyError:
        print("Warning: model not found. Using cl100k_base encoding.")
        encoding = tiktoken.get_encoding("cl100k_base")
    if model in {
        "gpt-3.5-turbo-0613",
        "gpt-3.5-turbo-16k-0613",
        "gpt-4-0314",
        "gpt-4-32k-0314",
        "gpt-4-0613",
        "gpt-4-32k-0613",
        }:
        tokens_per_message = 3
        tokens_per_name = 1
    elif model == "gpt-3.5-turbo-0301":
        tokens_per_message = 4  # every message follows <|start|>{role/name}\n{content}<|end|>\n
        tokens_per_name = -1  # if there's a name, the role is omitted
    elif "gpt-3.5-turbo" in model:
        print("Warning: gpt-3.5-turbo may update over time. Returning num tokens assuming gpt-3.5-turbo-0613.")
        return num_tokens_from_messages(messages, model="gpt-3.5-turbo-0613")
    elif "gpt-4" in model:
        print("Warning: gpt-4 may update over time. Returning num tokens assuming gpt-4-0613.")
        return num_tokens_from_messages(messages, model="gpt-4-0613")
    else:
        raise NotImplementedError(
            f"""num_tokens_from_messages() is not implemented for model {model}. See https://github.com/openai/openai-python/blob/main/chatml.md for information on how messages are converted to tokens."""
        )
    num_tokens = 0
    for message in messages:
        num_tokens += tokens_per_message
        for key, value in message.items():
            num_tokens += len(encoding.encode(value))
            if key == "name":
                num_tokens += tokens_per_name
    num_tokens += 3  # every reply is primed with <|start|>assistant<|message|>
    return num_tokens


In [16]:
# let's verify the function above matches the OpenAI API response

import openai

example_messages = [
    {
        "role": "system",
        "content": "You are a helpful, pattern-following assistant that translates corporate jargon into plain English.",
    },
    {
        "role": "system",
        "name": "example_user",
        "content": "New synergies will help drive top-line growth.",
    },
    {
        "role": "system",
        "name": "example_assistant",
        "content": "Things working well together will increase revenue.",
    },
    {
        "role": "system",
        "name": "example_user",
        "content": "Let's circle back when we have more bandwidth to touch base on opportunities for increased leverage.",
    },
    {
        "role": "system",
        "name": "example_assistant",
        "content": "Let's talk later when we're less busy about how to do better.",
    },
    {
        "role": "user",
        "content": "This late pivot means we don't have time to boil the ocean for the client deliverable.",
    },
]

for model in [
    "gpt-3.5-turbo-0301",
    "gpt-3.5-turbo-0613",
    "gpt-3.5-turbo",
    "gpt-4-0314",
    "gpt-4-0613",
    "gpt-4",
    ]:
    print(model)
    # example token count from the function defined above
    print(f"{num_tokens_from_messages(example_messages, model)} prompt tokens counted by num_tokens_from_messages().")
    # example token count from the OpenAI API
    response = client.chat.completions.create(
        model=model,
        messages=example_messages,
        temperature=0,
        max_tokens=1,  # we're only counting input tokens here, so let's not waste tokens on the output
    )
    
    print(f'{response.choices[0].message.content}')



gpt-3.5-turbo-0301
127 prompt tokens counted by num_tokens_from_messages().
This
gpt-3.5-turbo-0613
129 prompt tokens counted by num_tokens_from_messages().
This
gpt-3.5-turbo
129 prompt tokens counted by num_tokens_from_messages().
This
gpt-4-0314
129 prompt tokens counted by num_tokens_from_messages().
Changing
gpt-4-0613
129 prompt tokens counted by num_tokens_from_messages().
This
gpt-4
129 prompt tokens counted by num_tokens_from_messages().
This
